admin/installer/cloudformation.py

```diff
@@ -0,0 +1,293 @@
+            Output(
+                "AgentNode{}IP".format(i),
+                Description="Public IP of Agent Node for Flocker and Swarm.",
+                Value=GetAtt(ec2_instance, "PublicIp"),
+            )
+        ])
+
+    user_data += _sibling_lines(FLOCKER_CONFIGURATION_GETTER)
+    user_data += _sibling_lines(SIGNAL_CONFIG_COMPLETION)
+    ec2_instance.UserData = Base64(Join("", user_data))
+    template.add_resource(ec2_instance)
+
+# Client Node creation.
+client_instance = ec2.Instance(
+    CLIENT_NODE_NAME,
+    ImageId=FindInMap("RegionMap", Ref("AWS::Region"), "ClientAMI"),
+    InstanceType="m3.large",
+    KeyName=Ref(keyname_param),
+    SecurityGroups=[Ref(instance_sg)],
+    AvailabilityZone=zone,
+)
+wait_condition_handle = WaitConditionHandle(CLIENT_WAIT_HANDLE)
+template.add_resource(wait_condition_handle)
+wait_condition = WaitCondition(
+    CLIENT_WAIT_CONDITION,
+    Handle=Ref(wait_condition_handle),
+    Timeout="600",
+)
+template.add_resource(wait_condition)
+
+# Client Node {Flockerctl, Docker-compose} configuration.
+user_data = base_user_data[:]
+user_data += [
+    'wait_condition_handle="', Ref(wait_condition_handle), '"\n',
+    'node_number="{}"\n'.format("-1"),
+]
+user_data += _sibling_lines(S3_SETUP)
+user_data += _sibling_lines(CLIENT_SETUP)
+user_data += _sibling_lines(SIGNAL_CONFIG_COMPLETION)
+client_instance.UserData = Base64(Join("", user_data))
+client_instance.DependsOn = control_service_instance.name
+template.add_resource(client_instance)
+
+# List of Output fields upon successful creation of the stack.
+template.add_output([
+    Output(
+        "ClientNodeIP",
+        Description="Public IP address of the client node.",
+        Value=GetAtt(client_instance, "PublicIp"),
+    )
+])
+template.add_output(Output(
+    "ClientConfigDockerSwarmHost",
+    Value=Join("",
+               ["export DOCKER_HOST=tcp://",
+                GetAtt(control_service_instance, "PublicIp"), ":2376"]),
+    Description="Client config: Swarm Manager's DOCKER_HOST setting."
+))
+template.add_output(Output(
+    "ClientConfigDockerTLS",
+    Value="export DOCKER_TLS_VERIFY=1",
+    Description="Client config: Enable TLS client for Swarm."
+))
+base_url = "https://resources.console.aws.amazon.com/r/group#sharedgroup="
+parameters = {
+    "name": "%STACK_NAME%",
+    "regions": "all",
+    "resourceTypes": "all",
+    "tagFilters": [
+        {
+            "key": "aws:cloudformation:stack-name",
+            "values": [
+                "%STACK_NAME%"
+            ]
+        }
+    ]
+}
+parameter_string = json.dumps(parameters, separators=(',', ':'))
+variables = {
+    "STACK_NAME": Ref('AWS::StackName')
+}
+pattern = r'%([A-Z_]+)%'
+parts = re.split(pattern, parameter_string)
+iparts = iter(parts)
+new_parts = [base_url]
+for part in iparts:
+    new_parts.append(urllib.quote_plus(part))
+    key = next(iparts, None)
+    if key is not None:
+        new_parts.append(variables[key])
```

Comment:
From parameters to here is kind of hard to follow and a bit subtle. Looks like you need to generate an array  of (strings|refs) that can be concattenated into a url-encoded json blob after refs are resolved. I guess your guaranteed that AWS::StackName won't have any invalid URL characters, so this approach works. The hand-rolling of your own template language is a little non-obvious. Any chance you want to just put:

```
pattern = r'%([A-Z_]+)%'
 parts = re.split(pattern, parameter_string)
 iparts = iter(parts)
 new_parts = [base_url]
 for part in iparts:
     new_parts.append(urllib.quote_plus(part))
     key = next(iparts, None)
     if key is not None:
         new_parts.append(variables[key])
```

Into a helper function? Might make it a bit cleaner.